api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -150,6 +152,53 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_MATCH;
     }
 
+    @Override
+    public <T> Boolean isNaN(BoundReference<T> ref) {
+      Integer id = ref.fieldId();
+
+      if (nanCounts != null && nanCounts.containsKey(id) && nanCounts.get(id) == 0) {
+        return ROWS_CANNOT_MATCH;
+      }
+
+      // when there's no nanCounts information, but we already know the column only contains null,
+      // it's guaranteed that there's no NaN value
+      if (containsNullsOnly(id)) {
```

Comment:
Should we define a similar `containsNaNsOnly` method to use in `notNaN` and for a similar use in `isNull`?

Reply:
I didn't define `notNaN` originally as I could directly return `ROWS_CANNOT_MATCH` when both `nanCounts` and `valueCounts` contain this column but numbers don't match, without going into the next block of logic (of checking upper == lower == NaN and null count == 0); but this advantage no longer exists since that block needs to be removed. 

But I wasn't sure if we need it for `isNull`: currently in `isNull()` we are checking if `nullCounts == 0` to return `ROWS_CANNOT_MATCH`, and I guess the only chance where we rely on `containsNaNsOnly` to return `ROWS_CANNOT_MATCH` is `nullCounts` for this column doesn't exist but `nanCounts` does. I personally feel the chance of this happening would be small, do you think we will run into this case often?

Reply:
I agree that the `containsNaNsOnly` logic will not be very useful as Yan said, but I think it is also valuable to have that private method just for readability.

Then the question reduces to: do we need to consider the case that null value metrics do not exist but NaN metrics do. For now I think the answer is no, because in all metrics modes NaN and null counters either both exist or both not exist.

Reply:
Sounds good, I'll create a `containsNaNsOnly` for readability. Ryan, do you have comment on the other point?

Reply:
I agree with the reasoning. If we have NaN counts, then we should have null counts. No need to over-complicated the null logic with a check for when we don't have null counts but do have NaN counts. Good catch!